### PR TITLE
Upgrade to Mapbox GL–based Mapbox Static API

### LINKED ISSUE #9: Upgrade to Mapbox GL-based Mapbox Static API
This library uses the [classic Mapbox Static API](https://www.mapbox.com/api-documentation/#static-classic). We should upgrade it to the newer [Mapbox Static API](https://www.mapbox.com/api-documentation/#static) that’s based on Mapbox GL. The new API supports options like rotation and pitch, although it lacks support for markers and overlays.

/cc @friedbunny @bsudekum

### PR DESCRIPTION
**Pull request by 1ec5**
This PR adds support for the [Mapbox GL](https://wiki.openstreetmap.org/wiki/Mapbox_GL)–powered [Mapbox Static API](https://www.mapbox.com/api-documentation/#static) alongside the existing support for the classic Static API. In the process, the existing Swift files have been split up to be more manageable.

The `Snapshot` class and `Overlay` subclasses are largely untouched, a consequence of the fact that the new API tries to mimic the classic API’s overlay options. `SnapshotOptions` has been renamed `ClassicSnapshotOptions`. This PR takes advantage of the `SnapshotOptionsProtocol` protocol – meant to provide a common interface for `ClassicSnapshotOptions` and `MarkerOptions` – to add another conforming protocol, named `SnapshotOptions`, that exposes options for working with the new API.

`SnapshotOptions` is a bit more complex than `ClassicSnapshotOptions`. Instead of accepting separate `centerCoordinate` and `zoomLevel` options, it takes a single `camera` option of type `SnapshotCamera`. `SnapshotCamera` defines both a conventional `zoomLevel` property and an `altitude` property (for compatibility with the Mapbox iOS and macOS SDKs’ `MGLMapCamera` class); the developer can use either.

`SnapshotCamera` closely parallels the Mapbox iOS and macOS SDKs’ [MGLMapCamera](https://www.mapbox.com/ios-sdk/api/3.4.2/Classes/MGLMapCamera.html) class, in much the same way that MapKit’s MKMapSnapshotOptions and MKMapView classes both have `camera` properties of type MKMapCamera. This library can’t simply reuse MGLMapCamera because it lives in the Mapbox iOS and macOS SDKs, which are separate frameworks. (Perhaps we should move this entire library’s functionality into those SDKs to fix mapbox/mapbox-gl-native#1701?)

`GeoJSON(object:)` is once again throwable. #23 made it throwable, but #28 had to make it failable instead due to a type mismatch when bridging to Objective-C. As of Swift 3, [SE-0112](https://github.com/apple/swift-evolution/blob/master/proposals/0112-nserror-bridging.md) resolved the type mismatch, so the initializer can once again be throwable without breaking Objective-C support.

The classic snapshot tests have been rewritten to replace clever but untested URL parsing code in tests with much more straightforward string literals that can be verified easily. Now, if you want to know the URL that a particular `ClassicSnapshotOptions` produces, you can copy the stub’s `isPath()` call’s argument into a browser instead of having to run the test code. Test fixture images have been added, and the tests verify that an NSImage or UIImage can be created from each test fixture image produced by `Snapshot`.

To recap:

* [x] Reorganize Swift files
* [x] Rename `SnapshotOptions` to `ClassicSnapshotOptions`
* [x] Implement new `SnapshotOptions` based on GL-based API
* [x] Make `GeoJSON` initializer throwable
* [x] Make unit test stubs more literal
* [x] Make unit tests rely on image test fixtures
* [x] Add altitude (distance) as a camera option alongside zoom level, for consistency with MGLMapCamera
* [x] Add GL snapshot tests
* [x] Update example applications
* [x] Update playgrounds and readme
* [x] Update documentation comments
* [x] Update readme screenshots once server-side rendering bugs are fixed

Fixes #9.

/cc @frederoni @bsudekum @incanus @friedbunny

### CODE REVIEW

**Comment by 1ec5** [comment]
Test coverage is at 76.97%:

coverage

Most of the untested code is in convenience methods added for Objective-C compatibility.

**Comment by 1ec5** [comment]
The server-side rendering issues have been fixed, so the remaining work is no longer blocked.

File: MapboxStatic/Snapshot.swift [hunk 1]
```diff
@@ -66,315 +66,9 @@ public protocol SnapshotOptionsProtocol: NSObjectProtocol {
...
  
- If you use `Snapshot` to display a [vector tile set](https://www.mapbox.com/help/define-tileset/#vector-tilesets), the snapshot image will depict a wireframe representation of the tile set. To generate a static, styled image of a vector tile set, use the [vector Mapbox Static API](https://www.mapbox.com/api-documentation/?language=Swift#static).
+ The snapshot image can be used in an image view (`UIImage` on iOS and tvOS, `NSImage` on macOS, `WKImage` on watchOS). To add interactivity, use the `MGLMapView` class provided by the [Mapbox iOS SDK](https://www.mapbox.com/ios-sdk/) or [Mapbox macOS SDK](https://github.com/mapbox/mapbox-gl-native/tree/master/platform/macos/). (See the “[Custom raster style](https://www.mapbox.com/ios-sdk/examples/raster-styles/)” example to display a raster tile set in an `MGLMapView`.)
```
Comment (friedbunny) [inline MapboxStatic/Snapshot.swift]: > To add interactivity, use the `MGLMapView`...

This makes the two libraries sound as if one is necessary for the other or that they are somehow integrated. I think we should frame this more as, “For interactive maps, see our fully-featured SDKs...” and avoid mentioning classnames that impatient readers might expect to exist in this library.

I’m also unsure of the mention of the iOS SDK’s raster tiles example — it strikes me as a non-sequitor, given that that example deals in tilesets and not individual static images.

File: MapboxStatic/SnapshotOptions.swift [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    
+    /**
+     Radius of the Earth, measured in meters.
+     */
+    static let radiusOfEarth: CLLocationDistance = 6_378_137
+    
+    /**
+     Size of a tile.
+     */
+    static let tileSize = CGSize(width: 512, height: 512)
```
Comment (friedbunny) [inline MapboxStatic/SnapshotOptions.swift]: Strictly reading this code, it’s not clear that `tileSize` represents an internal implementation detail and not `size`, a thing the developer probably wants to customize.

Besides a little disambiguation, is there a source of truth for this that we can link to?

File: MapboxStatic/SnapshotOptions.swift [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+     - parameter zoomLevel: The zoom level of the snapshot.
+     - returns: A snapshot camera based on the given center coordinate and zoom level.
+     */
+    @objc(cameraLookingAtCenterCoordinate:zoomLevel:)
+    public static func camera(lookingAtCenter centerCoordinate: CLLocationCoordinate2D, zoomLevel: CGFloat) -> Self {
+        return self.init(lookingAtCenter: centerCoordinate, zoomLevel: zoomLevel)
+    }
+    
+    func string(size: CGSize) -> String {
+        assert(-90...90 ~= centerCoordinate.latitude, "Center latitude must be between −90° and 90°.")
```
Comment (friedbunny) [inline MapboxStatic/SnapshotOptions.swift]: Nice range comparison.

File: MapboxStatic/SnapshotOptions.swift [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        var zoomLevel: CGFloat = 0
+        if let level = self.zoomLevel {
+            zoomLevel = level
+        } else if let altitude = altitude {
+            zoomLevel = SnapshotCamera.zoomLevelForAltitude(altitude, pitch: pitch, latitude: centerCoordinate.latitude, size: size)
+        }
+        assert(0...20 ~= zoomLevel, "Zoom level must be between 0 and 20.")
+        components.append(Double(zoomLevel))
+        
+        if heading > 0 {
```
Comment (friedbunny) [inline MapboxStatic/SnapshotOptions.swift]: Does `heading` need a range check, or will >360 just keep wrapping?

File: MapboxStatic/SnapshotOptions.swift [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+     
+     The default value of this property matches the natural scale factor associated with the main screen. However, only images with a scale factor of 1.0 or 2.0 are ever returned by the Static API, so a scale factor of 1.0 of less results in a 1× (standard-resolution) image, while a scale factor greater than 1.0 results in a 2× (high-resolution or Retina) image.
+     */
+    open var scale: CGFloat = UIScreen.main.scale
+    #endif
+    
+    /**
+     A Boolean determining whether the resulting image includes the Mapbox logo.
+     
+     When shown, the Mapbox logo is located in the lower-left corner of the image. By default, this property is set to `true`.
```
Comment (friedbunny) [inline MapboxStatic/SnapshotOptions.swift]: Chances are quite high that the developer is required to include the logo, so linking https://www.mapbox.com/help/attribution/ here also makes sense.

File: README.md [hunk 6]
```diff
@@ -16,13 +16,13 @@ MapboxStatic.swift pairs well with [MapboxDirections.swift](https://github.com/m
 Specify the following dependency in your [Carthage](https://github.com/Carthage/Carthage/) Cartfile:
 
 ```sh
-github "Mapbox/MapboxStatic.swift" "~> 0.7"
+github "Mapbox/MapboxStatic.swift" "master"
```
Comment (friedbunny) [inline README.md]: Risky... is it necessary not to specify a version?

File: iOS.playground/Contents.swift [hunk 7]
```diff
@@ -5,30 +5,37 @@ import MapboxStatic
...
+ 
+ To generate a snapshot from a tile set, you’ll need a [map ID](https://www.mapbox.com/help/define-map-id/). You can either choose a [Mapbox-maintained tile set](https://www.mapbox.com/api-documentation/#maps) or upload your own to [Mapbox Studio](https://www.mapbox.com/studio/tilesets/).
+ 
+ You’ll also need an [access token](https://www.mapbox.com/help/define-access-token/) with the `styles:tiles` scope enabled in order to use this library. You can specify your access token inline or by setting the `MGLMapboxAccessToken` key in your application’s Info.plist file.
  */
 
-let mapIdentifiers = ["justin.tm2-basemap"]
+let styleURL = URL(string: "mapbox://styles/mapbox/streets-v9")!
+let mapIdentifiers = ["mapbox.satellite"]
 let accessToken = "pk.eyJ1IjoianVzdGluIiwiYSI6IlpDbUJLSUEifQ.4mG8vhelFMju6HpIY-Hi5A"
```
Comment (friedbunny) [inline iOS.playground/Contents.swift]: This token belongs to @incanus — it should be replaced with one belonging to the `mapbox` account, if we’re going to include one at all.

File: macOS.playground/Contents.swift [hunk 8]
```diff
@@ -5,30 +5,37 @@ import MapboxStatic
...
+ 
+ To generate a snapshot from a tileset, you’ll need a [map ID](https://www.mapbox.com/help/define-map-id/). You can either choose a [Mapbox-maintained tileset](https://www.mapbox.com/api-documentation/#maps) or upload your own to [Mapbox Studio](https://www.mapbox.com/studio/tilesets/).
+ 
+ You’ll also need an [access token](https://www.mapbox.com/help/define-access-token/) with the `styles:tiles` scope enabled in order to use this library. You can specify your access token inline or by setting the `MGLMapboxAccessToken` key in your application’s Info.plist file.
  */
 
-let mapIdentifiers = ["justin.tm2-basemap"]
+let styleURL = URL(string: "mapbox://styles/mapbox/streets-v9")!
+let mapIdentifiers = ["mapbox.satellite"]
 let accessToken = "pk.eyJ1IjoianVzdGluIiwiYSI6IlpDbUJLSUEifQ.4mG8vhelFMju6HpIY-Hi5A"
```
Comment (friedbunny) [inline macOS.playground/Contents.swift]: Ditto to this access token.

File: MapboxStatic/SnapshotOptions.swift (hunk 2 above)
Comment (1ec5) [inline MapboxStatic/SnapshotOptions.swift]: This is an private variable factored out for the purpose of keeping the altitude conversion code readable.

File: MapboxStatic/ClassicSnapshotOptions.swift [hunk 9]
```diff
@@ -0,0 +1,213 @@
+#if os(OSX)
+    import Cocoa
+#elseif os(watchOS)
+    import WatchKit
+#else
+    import UIKit
+#endif
+
+/**
+ A structure that determines what a snapshot depicts and how it is formatted. A classic snapshot is made by compositing one or more [tile sets](https://www.mapbox.com/help/define-tileset/) with optional overlays using the [classic Mapbox Static API](https://www.mapbox.com/api-documentation/?language=Swift#static-classic).
```
Comment (1ec5) [inline MapboxStatic/ClassicSnapshotOptions.swift]: Since Xcode automatically reformats the comment in the generated interface (whether you're coming from Swift or Objective-C code), I've been less concerned about documentation comment formatting here than in the gl-native codebase. But if you're concerned about readability for contributors to this codebase, we can address the line wrapping issue in a separate PR.

File: MapboxStatic/SnapshotOptions.swift (hunk 4 above)
Comment (1ec5) [inline MapboxStatic/SnapshotOptions.swift]: The API wraps it automatically.

File: README.md (hunk 6 above)
Comment (1ec5) [inline README.md]: For now. Whenever we rewrite the readme, we have to instruct the developer to get the master version so that the instructions are consistent. Otherwise we'd have to stage changes to the documentation, which would render them inconsistent with the rest of the repository in the meantime.

This shouldn't be a problem, because both CocoaPods and Carthage instruct developers to check in their lock files.

File: iOS.playground/Contents.swift (hunk 7 above)
Comment (1ec5) [inline iOS.playground/Contents.swift]: We do need to include the token if we want the playground to function as it does. We can replace it with a Mapbox corporate token, though.

File: MapboxStatic/ClassicSnapshotOptions.swift (hunk 9 above)
Comment (1ec5) [inline MapboxStatic/ClassicSnapshotOptions.swift]: We’d have to tread carefully, because the component that translates Swift into Objective-C headers (and thus reStructuredText into Doxygen) is a bit fragile. Objective-C bridging has broken in the past due to something as innocuous as a blank line between `- parameter` and `- note`.

Comment (friedbunny) [inline MapboxStatic/ClassicSnapshotOptions.swift]: Mostly the line length just made it slightly more painful to review.

File: MapboxStatic/SnapshotOptions.swift (hunk 5 above)
Comment (1ec5) [inline MapboxStatic/SnapshotOptions.swift]: Noted in 5ccc930a6990b6bca3d775fc23d2f48e7aed9937.

File: MapboxStatic/Snapshot.swift (hunk 1 above)
Comment (1ec5) [inline MapboxStatic/Snapshot.swift]: Clarified in 5ccc930a6990b6bca3d775fc23d2f48e7aed9937.

File: iOS.playground/Contents.swift (hunk 7 above)
Comment (1ec5) [inline iOS.playground/Contents.swift]: Replaced in 49ec1d0e26b1ae2b950a29ae81b383412755357a.

File: macOS.playground/Contents.swift (hunk 8 above)
Comment (1ec5) [inline macOS.playground/Contents.swift]: Dittoed in 49ec1d0e26b1ae2b950a29ae81b383412755357a.

**Comment by quicklywilliam** [comment]
Huzzah!
